Support hooks for MembershipType entity

mattwire opened the pull request:
Overview
----------------------------------------
Basic changes required to support pre/post hooks on MembershipType entity (see #11794).

Before
----------------------------------------
- pre/post hooks not triggered on edit/create.

After
----------------------------------------
- pre/post hooks triggered on edit/create.

eileenmcnaughton commented on CRM/Member/BAO/MembershipType.php:
```diff
@@ -128,6 +139,20 @@ public static function add(&$params, $ids = array()) {
       self::updateAllPriceFieldValue($id, $params);
     }
+
+    if (!empty($params['custom']) &&
```
@mattwire this should already be supported in the api layer  in 5.0  see 
https://github.com/civicrm/civicrm-dev-docs/commit/b0bf2bac2917b04ff617be32327be0eda3c80f16

We don't need to add via bao layer

eileenmcnaughton commented on CRM/Member/BAO/MembershipType.php:
```diff
@@ -91,11 +91,22 @@ public static function setIsActive($id, $is_active) {
+    if (empty($params['id'])) {
+      $params['id'] = CRM_Utils_Array::value('membershipType', $ids);
+    }
+    $id = CRM_Utils_Array::value('id', $params);
```
I think the tidiest way of expressing this part is the Phone BAO

    $hook = empty($params['id']) ? 'create' : 'edit';
    CRM_Utils_Hook::pre($hook, 'Phone', CRM_Utils_Array::value('id', $params), $params);

mattwire replied:
Agreed, updated to match that format

eileenmcnaughton commented on CRM/Member/BAO/MembershipType.php:
```diff
@@ -91,11 +91,22 @@ public static function setIsActive($id, $is_active) {
   public static function add(&$params, $ids = array()) {
-    $id = CRM_Utils_Array::value('id', $params, CRM_Utils_Array::value('membershipType', $ids));
+    if (empty($params['id'])) {
+      $params['id'] = CRM_Utils_Array::value('membershipType', $ids);
```
should we add a deprecation notice here?

mattwire replied:
Yes, added.

eileenmcnaughton commented on CRM/Core/SelectValues.php:
```diff
@@ -223,6 +223,7 @@ public static function customGroupExtends() {
       'ContributionRecur' => ts('Recurring Contributions'),
       'Group' => ts('Groups'),
       'Membership' => ts('Memberships'),
+      'MembershipType' => ts('Membership Types'),
```
not needed

mattwire replied:
@eileenmcnaughton The problem here is that you need to set 3 parameters (label, value, name) in cg_extend_objects but you cannot do this via the UI as name is not exposed.  So you are either limited to doing the DB change via an extension or editing the DB directly.  It may also be possible to do it via the API but it seems a lot less straightforward than just adding it here which allows the user to directly select "Membership Types" as a custom data entity when creating custom fields.
And we now have support in core for displaying those custom fields.

eileenmcnaughton replied:
@mattwire right - I think the intent IS that additional entities would be supported by adding values to that option group in an extension.

eileenmcnaughton commented on CRM/Core/BAO/CustomGroup.php:
```diff
@@ -1813,6 +1813,9 @@ public static function mapTableName($table) {
       case 'Membership':
         return 'civicrm_membership';
 
+      case 'MembershipType':
```
not needed

mattwire replied:
@eileenmcnaughton See comment above

eileenmcnaughton commented:
@mattwire the api already supports the custom data & we should switch to calling the api rather than doubling up on the BAO. The hook part we should add in the BAO create, as you have done. I'm in 2 minds about the ->find() addition because it imposes a performance cost on every save - regardless of whether hooks are in play. OTOH the hook itself can fetch extra data if required. (our code is inconsistent on this & as a low-traffic entity the performance is less of an issue but I think the precedent of leaving extra DB calls to the hook fn is better)

mattwire commented:
@eileenmcnaughton I've removed the ->find() per your comments.  It's not required so I'll not add it here.
I've also removed the reference to custom data in the BAO object as the API does indeed work now :-) Nice work with that!

eileenmcnaughton commented:
@mattwire I would merge this PR if it were just the BAO changes - I'm not sold on those 2 custom data changes as part of this

mattwire commented:
@eileenmcnaughton The PR is now just the BAO changes to support hooks for MembershipType. Assuming it passes tests would you mind having another look?

eileenmcnaughton commented:
unrelated fails

eileenmcnaughton commented:
Agree this just adds hooks & a little cleanup - merging
